Stable Diffusion via REST

adodge opened the pull request:
Hi!  Thanks for this project.  It's exactly the sort of thing I was looking for.  I made some nodes to hook into Automatic1111.

To get it running, you have to set up the service from here:  https://github.com/AUTOMATIC1111/stable-diffusion-webui and run it with `./webui.sh --api` to expose the REST API.  Here are some examples of it working:

![Screenshot from 2023-01-23 11-29-23](https://user-images.githubusercontent.com/551173/214135617-3ecfe8b6-e98a-4f74-8cf9-008c648dcd88.png)
![Screenshot from 2023-01-23 11-44-06](https://user-images.githubusercontent.com/551173/214135672-a5d437ec-ba4a-4130-9f8e-ad37e2e84ca3.png)

joeyballentine commented:
This is really cool! Thanks!

I'm not too sure how i feel about relying on a separate server to run this stuff though. It makes sense in that it's a lot easier to not have to integrate everything directly, but i feel like it has the potential to cause lots of issues and incompatibility given enough time. For example if automatic's SD repo updates and makes breaking changes, people would need to use an old version for this to work properly.

This would however be the perfect use case for a plugin once we have the plugin system implemented, but I'll defer to @rundevelopment on this one to see what he thinks.

joeyballentine commented:
Skimmed through the code. I like that you put this in a separate REST category to make it clear that it's interacting with external things and not anything built into chaiNNer.

If you don't have SD running, do these nodes still show up? What happens in that case?

adodge commented:
Thanks for taking a look!  Yes, all good points.

The nodes do show up, but obviously they report an error if you try to run them.  I could have it try to connect to the API when the nodes are imported.

joeyballentine commented:
We're currently talking about this, and we feel the benefits of adding this most likely outweigh any drawbacks. Are you in our Discord server? I think it would be easier to discuss this with you there, if you don't mind

RunDevelopment commented on backend/src/nodes/nodes/rest/img2img.py:
```diff
@@ -0,0 +1,99 @@
+        self.inputs = [
+            ImageInput(),
+            TextInput("Prompt", default="an astronaut riding a horse"),
+            TextInput("Negative Prompt", default=""),
```
This default value is invalid. Text inputs enforce a min length of 1 by default. If you want any input to be optional, use its `make_optional()` method. E.g. `TextInput("Negative Prompt").make_optional()`. If the user doesn't assign a value to an optional input, then the `run` of the node will get `None` for this parameter (be sure to denote this in the type hint of the parameter).

adodge replied:
Ah!  Okay, thank you.

RunDevelopment commented on backend/src/nodes/nodes/rest/img2img.py:
```diff
@@ -0,0 +1,99 @@
+            ImageInput(),
+            TextInput("Prompt", default="an astronaut riding a horse"),
+            TextInput("Negative Prompt", default=""),
+            NumberInput(
+                "Denoising Strength",
+                minimum=0,
+                default=0.75,
+                maximum=1,
+                controls_step=0.1,
+                precision=4,
+            ),
```
Generally, sliders are nicer for users because they visualize not only the current value but also the relation of the current value to the minimum and maximum.

![image](https://user-images.githubusercontent.com/20878432/214319177-81516431-fd58-4863-bca6-7a8530eb56b7.png)
```suggestion
            SliderInput(
                "Denoising Strength",
                minimum=0,
                default=0.75,
                maximum=1,
                slider_step=0.01,
                controls_step=0.1,
                precision=4,
            ),
```

RunDevelopment commented on backend/src/nodes/nodes/rest/img2img.py:
```diff
@@ -0,0 +1,99 @@
+                controls_step=0.1,
+                precision=4,
+            ),
+            NumberInput("Seed", minimum=0, default=42, maximum=4294967296),
```
We have a group (=wrappers around inputs that add functionality) for seeds. It adds a button that when pressed generates a random 6-digit number.

![image](https://user-images.githubusercontent.com/20878432/214319644-d1058dc7-aa21-42f4-8eae-e5fe4f455024.png)

(The `group` function is imported from `node_base`.)

```suggestion
            group("seed")(
                NumberInput("Seed", minimum=0, default=42, maximum=4294967296)
            ),
```

adodge replied:
Cool!  That is a lot nicer

RunDevelopment commented on backend/src/nodes/nodes/rest/img2img.py:
```diff
@@ -0,0 +1,99 @@
+                precision=4,
+            ),
+            NumberInput("Seed", minimum=0, default=42, maximum=4294967296),
+            NumberInput("Steps", minimum=1, default=20, maximum=150),
+            EnumInput(SamplerName, default_value=SamplerName.EULER),
+            NumberInput(
+                "CFG Scale",
```
Could Steps and CFG Scale also be slider inputs? The maximum values seem kinda weird, so I don't know if it makes sense.

Also, we also support sliders with log scales. So if you need that, we have it :)

adodge replied:
They are sliders in the Automatic1111 UI, so it would make sense here too.  The upper bounds are, I think, arbitrary but reasonable, and the same as what are in the other UI.

RunDevelopment commented on backend/src/nodes/nodes/rest/img2img.py:
```diff
@@ -0,0 +1,99 @@
+                precision=4,
+                controls_step=0.1,
+            ),
+            NumberInput("Width", minimum=64, default=512, maximum=2048),
+            NumberInput("Height", minimum=64, default=512, maximum=2048),
```
These are the output width and height of the image below, right? Does the REST service guarantee that it will return images with example those dimensions?

Can these be any number between 64 and 2048 or are only powers of 2 allowed? I remember playing with an SD upscaler, and it only correctly support multiples of 64.

adodge replied:
Yes.  I don't know if it explicitly guarantees the output size, but it does seem to either produce output of the given size or crash.  It looks like it works with any integer values, though I believe models work best on the dimensions they were trained on.  (Usually 512x512 or 768x768.)

In the Automatic UI, these are sliders from 64 to 2048.  I used another UI that had a dropdown menu for these with various canonical values.

Imitating the UI decisions of the thing we're calling seems like a good baseline.

RunDevelopment commented on backend/src/nodes/nodes/rest/img2img.py:
```diff
@@ -0,0 +1,99 @@
+            ),
+            NumberInput("Width", minimum=64, default=512, maximum=2048),
+            NumberInput("Height", minimum=64, default=512, maximum=2048),
+            TextInput("Model Checkpoint Override", default=""),
```
This should probably be an optional input too.

RunDevelopment commented on backend/src/nodes/impl/rest.py:
```diff
@@ -0,0 +1,95 @@
+)
+
+
+async def get_async(url) -> Dict:
+    async with aiohttp.ClientSession() as session:
+        async with session.get(url) as response:
+            return await response.json()
+
+
+async def post_async(url, json_data: Dict) -> Dict:
+    async with aiohttp.ClientSession() as session:
+        async with session.post(url, json=json_data) as response:
+            return await response.json()
```
Do these methods have to be async? The only reason you had to add `AsyncNodeBase` and everything around it is because of these methods.

We run nodes in a separate thread pool, so it should be fine to use a sync blocking API, I believe. I'm not a python expert, so I could be wrong here.

adodge replied:
I'll investigate...  It seemed necessary at the time, but maybe I didn't have the full context.

adodge replied:
Oh yeah.  Absolutely unnecessary.  Works fine with `requests` instead of `aiohttp`.

RunDevelopment commented on backend/src/nodes/impl/rest.py:
```diff
@@ -0,0 +1,95 @@
+            return await response.json()
+
+
+STABLE_DIFFUSION_OPTIONS = asyncio.run(get_async(STABLE_DIFFUSION_OPTIONS_URL))
```
This doesn't seem to be used anywhere. What's its purpose?

adodge replied:
Currently it's purpose is to make a call to the API at import time that will fail if it's not available, causing the nodes to be hidden.  I'll make a comment.

RunDevelopment reviewed (changes requested):
Thank you for this @adodge! Great work!

I left you a few comments with questions and suggestions.

joeyballentine commented on backend/src/nodes/nodes/rest/__init__.py:
```diff
@@ -0,0 +1,8 @@
+from ...category import Category
+
+category = Category(
+    name="REST",
+    description="Interact with external components via REST",
```
"Stable Diffusion (External)" maybe?

adodge replied:
I do like the idea of an external category with the specific service as a subcategory.  We could call it "Automatic1111" instead of "Stable Diffusion", just to make that relationship more explicit.

joeyballentine replied:
Or maybe "Automatic1111 Stable Diffusion"? I feel like SD should be referenced somewhere

adodge replied:
How about: category "Stable Diffusion (External)" subcategory "Automatic1111"

adodge replied:
I guess a question is whether there's a connection between categories and dependencies.  Right now, it looks like it's 1-to-1.  Installing PyTorch unlocks the entire PyTorch category.

joeyballentine replied:
Yeah I'd say go with that. Makes enough sense.

adodge commented on backend/src/nodes/impl/rest.py:
```diff
@@ -0,0 +1,95 @@
+from .image_utils import normalize
+from ..utils.utils import get_h_w_c
+
+STABLE_DIFFUSION_HOST = os.environ.get("STABLE_DIFFUSION_HOST", "127.0.0.1")
```
Good point.  How about `external_stable_diffusion.py`?

adodge commented:
Implemented the feedback given.  Here are some new screenshots:

![image](https://user-images.githubusercontent.com/551173/214441867-1857c2d5-9af2-4dba-8d5d-ba781d8a426d.png)
![image](https://user-images.githubusercontent.com/551173/214441910-313cc8dd-b63a-4cd0-9792-c1503b5f43df.png)

joeyballentine commented:
The results of this all still get saved by automatic's repo right? Like there's no chance someone would generate something cool but forget to save it and never be able to get it back?

adodge commented:
Weirdly it doesn't save, and I don't see an option in the API to do so.  Would be good to have an easy way to do something similar on the ChaiNNer side.  Maybe a node that outputs the current time as text, that could be used with the save node.

There's also the problem of saving metadata.  The API does give a JSON blob with metadata, so that could be an output from the SD node, and we could have a "Save Text" node.

adodge commented:
Here's a possible implementation of the saving:

![image](https://user-images.githubusercontent.com/551173/214452921-50bdbae3-d9af-41e9-8238-9c51ef9290ab.png)
![image](https://user-images.githubusercontent.com/551173/214452617-075124dd-7f78-48ea-9c4a-c66ee40f6a0b.png)

It's kind of a lot of spaghetti to do something basic.  Maybe there could be one node that does it all.

RunDevelopment commented on requirements.txt:
```diff
@@ -17,6 +17,7 @@ pylint
 pyright==1.1.280
 pytest
 pywin32 ; sys_platform=="win32"
+requests
```
We also need to declare this dependency in `dependencies.ts` (the file chainner uses to manage python deps inside the GUI). 

@joeyballentine Should this be a required dep or an optional dep? `requests` a pretty commonly used (and small) library, so I wouldn't mind making it a required dep.

joeyballentine replied:
I'm fine with it being required

RunDevelopment commented on backend/src/nodes/nodes/external_stable_diffusion/img2img.py:
```diff
@@ -0,0 +1,102 @@
+        ]
+
+        self.category = ExternalStableDiffusionCategory
+        self.name = "Image-to-Image"
```
Could we call this "Image To Image"? We don't use kebab case anywhere else. Same for text to image.

RunDevelopment commented on backend/src/nodes/impl/external_stable_diffusion.py:
```diff
@@ -0,0 +1,96 @@
+    return response.json()
+
+
+# Call the API at import time
+# If this fails (because the API isn't up) it will raise an exception and the
+# nodes importing this file won't be registered
+STABLE_DIFFUSION_OPTIONS = get(STABLE_DIFFUSION_OPTIONS_URL)
```
I wonder if this is a good idea. When I initially checked out this PR, the SD nodes didn't even show up because it couldn't connect to the options URL for whatever reason. I'm also not sure whether connecting to this URL once is a reliable indicator for the availability of the service. The nodes have to handle network errors anyway, since the device might disconnect for the network (laptop).

Hiding nodes based on factors outside the user's machine also presents usability issues. Imagine a user creating a chain with SD yesterday. Today the SD service has maintenance and so it refuses connections. Suddenly, the chain the user create but yesterday fails to load as chainner doesn't recognize the SD nodes anymore.

Yeah, I think having an error explaining that the service is unavailable at the moment would be preferable.

joeyballentine replied:
It's not outside the user's machine though, this is a local server it's connecting to.

And my biggest thing was just not confusing people. We hide pytorch nodes if pytorch isn't installed, i thought this should be similar.

RunDevelopment replied:
Oh, you're right. I didn't even look at the host address and just assumed it was some remote server.

RunDevelopment commented on backend/src/nodes/nodes/external_stable_diffusion/img2img.py:
```diff
@@ -0,0 +1,102 @@
+            TextInput("Model Checkpoint Override").make_optional(),
+        ]
+        self.outputs = [
+            LargeImageOutput(),
```
`LargeImageOutput` implicitly have side effects (the image is shown to the user). Please use a regular `ImageOutput`. Same for text to image.

adodge commented:
Made a few more changes in response to feedback:

- Added requests library to requirements.ts
- Removed kebab case in favor of "Text to Image"
- Changed LargeImageOutput to ImageOutput

RunDevelopment commented on backend/src/nodes/nodes/external_stable_diffusion/img2img.py:
```diff
@@ -0,0 +1,102 @@
+            TextInput("Model Checkpoint Override").make_optional(),
+        ]
+        self.outputs = [
+            ImageOutput(),
```
We talked about the Width and Height inputs before and how the REST service seems to guarantee that the returned images have those dimensions. If we actually guarantee that the returned images have those dimensions (e.g. by checking and throwing an error if they aren't the right size), then we can denote this in the output type here.

Chainner has a custom type system that allows us to verify chains and to predict output types. E.g. when you see chainner knowing the exact size of all output images in [this screenshot](https://github.com/chaiNNer-org/chaiNNer/blob/main/src/public/screenshot.png), that's the type system at work. 

Assuming that the Width and Height inputs are correct, we can denotes in the `ImageOutput` like this:
```py
ImageOutput(image_type="Image { width: Input8, height: Input9 }")
```

`Input8` and `Input9` are the Width and Height inputs. All inputs are assigned an id based on their position (starting with 0) and those to inputs have the ids 8 and 9 respectively.

But we can potentially do even better. Do we also know the number of channels of the returned image? E.g. if given a grayscale, RGB, or RGBA image, what type of image will be returned? If we do, then we can denote this in the type as well.

adodge replied:
Very cool!

It looks like `Image to Image` will always return an RGB image of the size you specify, even if it's given a greyscale or RGBA input, or an input with a different size.  `Text to Image` is the same.

I added these notes to the ImageOutputs.

RunDevelopment reviewed (commented):
One last thing for the Img2Img and Tex2Img nodes.

RunDevelopment commented on backend/src/nodes/nodes/external_stable_diffusion/img2img.py:
```diff
@@ -0,0 +1,104 @@
+        ]
+        self.outputs = [
+            ImageOutput(
+                image_type="Image {width: Input8, height: Input9, channels: 3}"
```
Using `channels: x` inside the type expression does work, but we have something better. `ImageOutput` also has a [`channels` parameter](https://github.com/chaiNNer-org/chaiNNer/blob/main/backend/src/nodes/properties/outputs/numpy_outputs.py#L42). This parameter not only adds the channel number to the type, it also enforces it at runtime. So if the API returns something that is not an RGB image, the user will see an error saying that the node is implemented incorrectly. This is important because chainner uses type information to verify the correctness of chains, so the type information has to be correct.

Same for text to image.

```suggestion
                image_type="Image { width: Input8, height: Input9 }",
                channels=3,
```

RunDevelopment commented on backend/src/nodes/nodes/external_stable_diffusion/img2img.py:
```diff
@@ -0,0 +1,104 @@
+                "sd_model_checkpoint"
+            ] = sd_model_checkpoint
+        response = post(url=STABLE_DIFFUSION_IMG2IMG_URL, json_data=request_data)
+        return decode_base64_image(response["images"][0])
```
As with channels, we should also have a runtime check here to verify the size of the returned image.

Same for text to image.
```suggestion
        result = decode_base64_image(response["images"][0])
        h, w, _ = get_h_w_c(result)
        assert (w, h) == (width, height), f"Expected the returned image to be {width}x{height}px but found {w}x{h}px instead"
        return result
```

adodge replied:
Cool!  Yes, makes sense.

> Sorry about this.

No worries!  I appreciate you taking the time to review it.

adodge replied:
Actually, this has raised a weird issue that I missed.  The API does not in fact always return an image of the size you request.  I thought I'd tested this, but I guess I happened to only try multiple of 8, because actually it seems to round down.

If you ask for 513x513, it will return 512x512.  If you ask for 511x511 it will return 504x504.

I added `control_step` and `slider_step` arguments to the SliderInputs, but it doesn't stop someone from manually setting an invalid size.

https://github.com/chaiNNer-org/chaiNNer/blob/059bb95a1fef73b30935a25dd8624dcc360fa6cf/backend/src/nodes/nodes/external_stable_diffusion/img2img.py#L60-L67

For now I added this runtime error, but I think it'd be better if we could predict the output size instead.  Is it possible, I wonder, to specify this "round down" behavior in the image type of the ImageOutput?  I get lost when I try to find out where things are happening in the typescript code.

https://github.com/chaiNNer-org/chaiNNer/blob/059bb95a1fef73b30935a25dd8624dcc360fa6cf/backend/src/nodes/nodes/external_stable_diffusion/img2img.py#L101-L104

RunDevelopment replied:
> For now I added this runtime error, but I think it'd be better if we could predict the output size instead. Is it possible, I wonder, to specify this "round down" behavior in the image type of the ImageOutput? I get lost when I try to find out where things are happening in the typescript code.

Navi docs are [here](https://github.com/chaiNNer-org/Navi/blob/main/navi.md#built-in-functions), but it seems like I forgot to document the `floor` and `ceil` functions. I'll fix that soon.
And yes, there is a lot of TS code for the type system. It's a whole little programming language :)

Anyway, yes, we can adjust the output type to do rounding to get rid of the runtime error. Since we have a `floor` function, we can do `int & floor(InputX / 8) * 8` to round down to the nearest multiple of 8.

If you want to, you can even extract this into a function:
```
def nearest_valid(n: number) = int & floor(n / 8) * 8;
Image { 
    width: nearest_valid(InputX),
    height: nearest_valid(InputY)
}
```

adodge replied:
Oh, awesome!  I didn't realize there was documentation!  Thank you.  I've put that in and it seems to be working.

RunDevelopment reviewed (changes requested):
Sorry about this. Last last last change request.

adodge commented:
@joeyballentine Oops.  Done!

joeyballentine reviewed (approved):
Awesome. Thanks again for this
